### ISSUE TITLE
Support state with different type of data

### ISSUE THREAD

**Issue by thangngoc89** [Astrocoders/reform#23]
Example:

```reason
module FormParams = {
  type state = {
    name: string,
    age: int,
  };
  type fields = [ | `name |  `age ];
  let lens = [
    (`name, s => s.name, (s, name) => {...s, name}),
    (
      `age,
      s => s.age,
      (s, age) => {...s, age}
    ),
  ];
};
```

My current work around is make a variant like this: 

```
type formValue = | Int(int) | String(string); 
```

Is this the right approach? Could reform support this use case?

**Comment by grsabreu** [comment]
Unfortunately indeed only `string` values are handled now.
My plan is to implement this https://github.com/Astrocoders/reform/issues/17

**Comment by thangngoc89** [comment]
@grsabreu ah. Now I understand what is that issue about ;)

**Comment by grsabreu** [comment]
yeah haha
I did it just for `string` for now because I was afraid of making it too boilerplate-y meanwhile I was thinking in a way of doing it. But I think the best way is indeed to make a variant out of it. Unfortunately it'll oblige the user to unwrap the values from the variant in the end before using it :/

The values are defined in https://github.com/Astrocoders/reform/blob/master/re/ReForm.re#L48

**Comment by thangngoc89** [comment]
@grsabreu yep. I've read through your code and with my current limited knowledge of OCaml, I honestly have no idea how to improve this. But I really love the library, handling form in a type safe manner. Maybe asking this on Discord ?

**Comment by grsabreu** [comment]
It's a good idea!
I think this will require some major change in thinking about keeping the form state out of a regular record.
I'm still polishing up the concept in my mind still and I think I'm going to open a discussion soon for help in Discord or the new forums

**Comment by grsabreu** [comment]
I'm thinking here in provide a hardcoded way to allow the basic scalar types in ReForm. What would make the usage less flexible but more succinct and I think covers 95% of the use cases. So the usage would be

```re
module FormParams = {
  type state = {
    name: string,
    age: int,
    favoriteColors: list(string),
  };
  type fields = [ | `name |  `age ];
  let lens = [
    String(`name, s => s.name, (s, name) => {...s, name}),
    Int(
      `age,
      s => s.age,
      (s, age) => {...s, age}
    ),
  ];
};
```

**Comment by guilhermedecampo** [comment]
How favoriteColors will work in this example @grsabreu? 🤔

**Comment by grsabreu** [comment]
I've been playing around with some GADTs concepts to bring a more advanced and flexible API to ReForm.
You can check it here

https://sketch.sh/s/ab4xNR6MC0QRYv3a1Q5boX/

**Comment by grsabreu** [comment]
Inspired by https://reasonml.chat/t/packing-gadt-constructors-in-iterable-data-structure/1120/5

**Comment by Anahkiasen** [comment]
It should maybe be made clearer that int and such is not accepted, I tried to model what I thought was a basic state and I got this really confusing error:

```reason
module RunFormParams = {
  type state = {time: int};
  type fields = [ | `time];

  let lens = [(`time, s => s.time, (_s, time) => {time: time})];
};
```

```
  41 │ };
  42 │
  43 │ module RunForm = ReForm.Create(RunFormParams);
  44 │
  45 │ let component = ReasonReact.reducerComponent("App");

  Signature mismatch:
  ...
  Values do not match:
    let lens: list(([> `time ], state => int, ('a, int) => state))
  is not included in
    let lens:
      list((fields, state => ReForm.Value.t,
           (state, ReForm.Value.t) => state))
```

Which I tried to Google to no avail until I came here :/

**Comment by grsabreu** [comment]
Hey @Anahkiasen , sorry that the documentation was not clear about it.
But we indeed have a type limitation for strings only, which is being addressed with #64 .

Meanwhile what you can do is to store the int as a string and then convert with int_of_string whenever you need to consume it

**Comment by Anahkiasen** [comment]
Yeah it's no biggie, was mostly confusion on my end as it's the first ReasonML form library I use so I already wasn't sure how Reason handled number inputs and such. Rest of the lib is pretty great so far 👍
